libs/grok/src/main/java/org/elasticsearch/grok/Grok.java

```diff
@@ -162,49 +164,63 @@ public String groupMatch(String name, Region region, String pattern) {
      *
      * @return named regex expression
      */
-    public String toRegex(String grokPattern) {
-        byte[] grokPatternBytes = grokPattern.getBytes(StandardCharsets.UTF_8);
-        Matcher matcher = GROK_PATTERN_REGEX.matcher(grokPatternBytes);
-
-        int result;
-        try {
-            matcherWatchdog.register(matcher);
-            result = matcher.search(0, grokPatternBytes.length, Option.NONE);
-        } finally {
-            matcherWatchdog.unregister(matcher);
-        }
-        if (result >= 0) {
-            Region region = matcher.getEagerRegion();
-            String namedPatternRef = groupMatch(NAME_GROUP, region, grokPattern);
-            String subName = groupMatch(SUBNAME_GROUP, region, grokPattern);
-            // TODO(tal): Support definitions
-            @SuppressWarnings("unused")
-            String definition = groupMatch(DEFINITION_GROUP, region, grokPattern);
-            String patternName = groupMatch(PATTERN_GROUP, region, grokPattern);
+    public String toRegex(String grokPattern, int recurseCount) {
```

Comment:
I think we should remove recursion here altogether, that way we can avoid stack overflows without using additional threads and context switching. As this is tail recursion it is fairly easy and straightforward:
```java
public String toRegex(String grokPattern) {
    StringBuilder res= new StringBuilder();
    while (true) {
        byte[] grokPatternBytes = grokPattern.getBytes(StandardCharsets.UTF_8);
        Matcher matcher = GROK_PATTERN_REGEX.matcher(grokPatternBytes);

        int result;
        try {
            matcherWatchdog.register(matcher);
            result = matcher.search(0, grokPatternBytes.length, Option.NONE);
        } finally {
            matcherWatchdog.unregister(matcher);
        }

        if (result < 0) {
            return res.append(grokPattern).toString();
        }

        Region region = matcher.getEagerRegion();
        String namedPatternRef = groupMatch(NAME_GROUP, region, grokPattern);
        String subName = groupMatch(SUBNAME_GROUP, region, grokPattern);// TODO(tal): Support definitions
        @SuppressWarnings("unused")
        String definition = groupMatch(DEFINITION_GROUP, region, grokPattern);
        String patternName = groupMatch(PATTERN_GROUP, region, grokPattern);
        String pattern = patternBank.get(patternName);
        if (pattern == null) {
            throw new IllegalArgumentException("Unable to find pattern [" + patternName + "] in Grok's pattern dictionary");
        }
        if (pattern.contains("%{" + patternName + "}") || pattern.contains("%{" + patternName + ":")) {
            throw new IllegalArgumentException("circular reference in pattern back [" + patternName + "]");
        }
        String grokPart;
        if (namedCaptures && subName != null) {
            grokPart = String.format(Locale.US, "(?<%s>%s)", namedPatternRef, pattern);
        } else if (namedCaptures) {
            grokPart = String.format(Locale.US, "(?:%s)", pattern);
        } else {
            grokPart = String.format(Locale.US, "(?<%s>%s)", patternName + "_" + result, pattern);
        }
        String start = new String(grokPatternBytes, 0, result, StandardCharsets.UTF_8);
        String rest = new String(grokPatternBytes, region.end[0], grokPatternBytes.length - region.end[0], StandardCharsets.UTF_8);
        grokPattern = grokPart+rest;
        res.append(start);
    }
}
```

Reply:
thanks @probakowski I will switch the implementation based on this.